Shipping Labels: Update tabs for shipments with purchased labels

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15440 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR implements a few changes to support showing existing shipments on the split shipments screen:
- Updated the order details screen to display the Create shipping label button even when a label is created. This is a temporary solution until #15375 is done. This change is also hidden under the feature flag to ensure that the behavior of the legacy flow stays the same.
- Updated the shipping label creation form to only load shipment configs and show the split shipments entry point if there are more than 1 item in the order.
- Updated the split shipments screen to display existing shipments in the loaded configs.
- Updated the shipment tabs when a shipment's label has already been purchased.
- Disabled removing purchased shipments.
- Updated the logic of merging all unfulfilled shipments to exclude purchased ones.

Note: the design found for the purchased shipment shows the checkbox disabled, but I think hiding the checkbox is simpler so I'm keeping that behavior.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Testing the Create shipping label button on the order details screen
1. Log in to a test store with Woo Shipping set up.
2. On the web, open a complete order with more than 1 physical product with different quantities - split shipments and purchase the label for one of the shipments.
3. Open the same order on the app and confirm that the Create shipping label button is available.
4. Switch to a different store with only the legacy WCShip plugin set up.
5. Open an order with a purchased shipping label and confirm that the Create shipping label button is unavailable.

### Testing the split shipments entry point
1. Log in to a test store with Woo Shipping set up.
2. On the app, open a complete order with more than 1 physical product with different quantities.
3. Select Create shipping label and confirm that the Split shipment entry point is displayed.
4. Navigate back to the order list and select an order with only 1 physical product with 1 item.
5. Select Create shipping label and confirm that the Split shipment entry point is not present.

### Testing existing shipments
1. Log in to a test store with Woo Shipping set up.
2. On the web, open a complete order with more than 1 physical product with different quantities - split shipments and purchase the label for one of the shipments.
3. Open the same order on the app and select Create shipping label > Split shipments.
4. Confirm that the correct shipments are displayed.
5. Confirm that the shipment with the purchased label is displayed with a check icon on the tab bar, and the items are not selectable.
6. Tap the ellipsis button on the tab bar and confirm that the purchased shipment is not enabled as an option to remove.
7. Select to remove any other shipments and confirm in the bottom sheet that the option to merge into the purchased shipment is disabled.
8. Tap the ellipsis button again and select Merge all unfulfilled. 
9. Confirm that only unfulfilled shipments are merged.


## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Tested and confirmed the above test cases in simulator iPhone 16 Pro iOS 18.2.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->



https://github.com/user-attachments/assets/7a5cb3bd-7a9a-4fe4-a11a-9fb52c0af108




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Add options to customize tab icons

itsmeichigo committed:
Add icon for purchased shipments

itsmeichigo committed:
Add text for purchased shipment

itsmeichigo committed:
Add isSelectable to CollapsibleShipmentItemCardViewModel initializer

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Enable shipping label creation button for orders with purchased labels (temporary solution)

itsmeichigo committed:
Create shipments based on configs

itsmeichigo committed:
Skip fetching configs if order has only one item

itsmeichigo committed:
Add test for splitting shipments entry point

itsmeichigo committed:
Add tests for shipment creation

itsmeichigo committed:
Disable the option to remove and merge purchased shipments

itsmeichigo committed:
Update logic for merging unfulfilled shipments

itsmeichigo committed:
Add test for merging unfulfilled shipments when there's a purchased shipment

itsmeichigo committed:
Add workaround when shipment ID is returned as a number instead of string

itsmeichigo commented on Networking/Networking/Model/ShippingLabel/ShippingLabelPurchase.swift:
```diff
@@ -104,7 +104,9 @@ extension ShippingLabelPurchase: Decodable {
         let productNames = try container.decode([String].self, forKey: .productNames)
 
-        let shipmentID = try container.decodeIfPresent(String.self, forKey: .shipmentID)
+        let shipmentID = container.failsafeDecodeIfPresent(targetType: String.self,
+                                                           forKey: .shipmentID,
+                                                           alternativeTypes: [.integer(transform: { String($0) })])
```
While testing again I saw that the shipment ID was somehow returned as a number today. This is a workaround to transform the number to our expected string type.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Split Shipments/WooShippingSplitShipmentsDetailView.swift:
```diff
@@ -25,6 +25,23 @@ struct WooShippingSplitShipmentsDetailView: View {
 
                 ScrollView {
                     VStack(alignment: .leading, spacing: Layout.contentPadding) {
+                        if viewModel.currentShipment.isPurchased {
+                            VStack {
+                                Text(Localization.PurchasedShipment.title)
+                                    .bold()
+                                    .frame(maxWidth: .infinity, alignment: .leading)
+                                Text(Localization.PurchasedShipment.subtitle)
+                                    .frame(maxWidth: .infinity, alignment: .leading)
+                            }
+                            .multilineTextAlignment(.leading)
+                            .foregroundStyle(Layout.green)
+                            .padding(Layout.contentPadding)
+                            .background(
+                                Layout.greenBackground
+                                    .clipShape(RoundedRectangle(cornerRadius: Layout.cornerRadius))
+                            )
+                        }
```
super-nit: Consider moving this into a computed property to keep the main `body` less crowded.  Perhaps named as `shipmentPurchasedMessageView`.

itsmeichigo replied:
Done in 4e310d25a4bcea35c913df21017735e2bacac848.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShippingCreateLabelsViewModel.swift:
```diff
@@ -275,8 +275,12 @@ final class WooShippingCreateLabelsViewModel: ObservableObject {
 
-            group.addTask {
-                await self.loadShipmentsInfo()
+            let totalOrderItems = order.items.map(\.quantity).reduce(0, +)
+            if totalOrderItems > 1 {
+                // Only fetch shipments info if there are more than one order items.
+                group.addTask {
+                    await self.loadShipmentsInfo()
+                }
```
Suggestion: We could move this `totalOrderItems` check logic into `loadShipmentsInfo` to keep the shipments-related logic inside one method.

itsmeichigo replied:
I plan to remove this check when working on #15309, so I'll keep this for now.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Split Shipments/WooShippingSplitShipmentsViewModel.swift:
```diff
@@ -78,8 +78,10 @@ final class WooShippingSplitShipmentsViewModel: ObservableObject {
         shipments.enumerated().map { (index, item) in
-            TopTabItem(name: String.localizedStringWithFormat(Localization.shipmentFormat, index + 1),
-                       content: { EmptyView() })
+            let purchasedIcon = UIImage(systemName: "checkmark.circle.fill")?.withRenderingMode(.alwaysTemplate)
+            return TopTabItem(name: String.localizedStringWithFormat(Localization.shipmentFormat, index + 1),
+                              icon: item.isPurchased ? purchasedIcon : nil,
+                              content: { EmptyView() })
```
nit: We could initialise the icon outside of the `map` to avoid creating the image for every shipment. 

```suggestion
        let purchasedIcon = UIImage(systemName: "checkmark.circle.fill")?.withRenderingMode(.alwaysTemplate)
        return shipments.enumerated().map { (index, item) in
            TopTabItem(name: String.localizedStringWithFormat(Localization.shipmentFormat, index + 1),
                       icon: item.isPurchased ? purchasedIcon : nil,
                       content: { EmptyView() })
        }
```

itsmeichigo replied:
Done in 8a33bbba27d25ec3252855e802632b998b1b004e.

selanthiraiyan reviewed (approved):
The code looks great and the given test cases all passed. :shipit: 

---

For an order with products with more than one quantities (with sub items), I noticed that after the shipments are saved from mobile the web isn't able to properly parse the saved information. i.e. The split shipments interface in web isn't working properly after the shipment is split from the mobile app. 

https://github.com/user-attachments/assets/568ba81f-99dc-4d48-bdd6-c2b4dac36938

This could be either an issue with how we save the shipments or how the web parses the shipments info. I think we should log a separate issue for this and investigate it. 

I have logged an issue here. https://github.com/woocommerce/woocommerce-ios/issues/15469

itsmeichigo committed:
Move fulfilled message to a separate view

itsmeichigo committed:
Move purchased icon to a constant

itsmeichigo commented:
Thank you @selanthiraiyan for the review! 

> For an order with products with more than one quantities (with sub items), I noticed that after the shipments are saved from mobile the web isn't able to properly parse the saved information. i.e. The split shipments interface in web isn't working properly after the shipment is split from the mobile app.

Good catch!

I have addressed the suggestions in subsequent commits, enabling auto-merge for now.
